File: 2020/julia/src/Day23.jl
module Day23
include("Aoc.jl")
using .Aoc

labels = [parse(Int64, c) for c in Aoc.input_line(23)]

mutable struct Cup
    label::Int64
    next::Union{Nothing,Cup}
    Cup(label::Int64) = new(label, nothing)
end

pick_label(label::Int64, count::Int64) = label > 1 ? label - 1 : count

function play(labels::Array{Int64}, rounds::Int64)
    count = length(labels)
    cups = [Cup(l) for l in labels]
    cups_map = Dict{Int64,Cup}()

    for (i,c) in enumerate(cups)
        ni = i == count ? 1 : i + 1
        c.next = cups[ni]
        cups_map[c.label] = c
    end

    current = cups[1]

    for i in 1:rounds
        one = current.next
        two = one.next
        three = two.next

        after_taken = three.next
        current.next = after_taken

        dest_label = current.label
        dest_label = pick_label(dest_label, count)
        while dest_label == one.label || dest_label == two.label || dest_label == three.label
            dest_label = pick_label(dest_label, count)
        end

        destination = cups_map[dest_label]
        oldNext = destination.next
        destination.next = one
        three.next = oldNext

        current = current.next
    end

    cups_map[1]
end

function part1()
    cup = play(labels, 100)

    result = ""
    for i in 1:8
        cup = cup.next
        result = result * string(cup.label)
    end

    result
end

function part2()
    cup = play(vcat(labels, 10:1_000_000), 10_000_000)
    return cup.next.label * cup.next.next.label
end
end

File: 2020/php/src/Aoc/Days/Day23.php
<?php

namespace Ppx17\Aoc2020\Aoc\Days;

use Ppx17\Aoc2020\Aoc\Days\Common\Cup;

class Day23 extends AbstractDay
{
    private array $cups;

    public function dayNumber(): int
    {
        return 23;
    }

    public function setUp(): void
    {
        ini_set('memory_limit', '256M');
        $this->cups = array_map(fn($x) => (int)$x, str_split($this->getInputTrimmed()));
    }

    public function part1(): string
    {
        $cup = $this->play($this->cups, 100);

        $result = '';
        for ($i = 0; $i < 8; $i++) {
            $cup = $cup->next;
            $result .= $cup->label;
        }

        return $result;
    }

    public function part2(): string
    {
        $cups = array_merge($this->cups, range(max($this->cups) + 1, 1_000_000));
        $firstCup = $this->play($cups, 10_000_000);

        return $firstCup->next->label * $firstCup->next->next->label;
    }

    private function play(array $labels, int $rounds): Cup
    {
        $count = count($labels);
        $cups = array_map(fn($x) => new Cup($x), $labels);
        $cupsByLabel = [];
        foreach ($cups as $index => $cup) {
            $ni = $index + 1;
            if ($ni === $count) {
                $ni = 0;
            }
            $cup->next = $cups[$ni];
            $cupsByLabel[$cup->label] = $cup;
        }

        $current = $cups[0];
        for ($i = 0; $i < $rounds; $i++) {
            // Take the three following cups
            $one = $current->next;
            $two = $one->next;
            $three = $two->next;

            $afterTaken = $three->next;
            $current->next = $afterTaken;

            $destLabel = $current->label;
            do {
                $destLabel = ($destLabel > 1) ? $destLabel - 1 : $count;
            } while ($destLabel === $one->label || $destLabel === $two->label || $destLabel === $three->label);

            $destination = $cupsByLabel[$destLabel];
            $oldNext = $destination->next;

            // Link the taken sequence back in
            $destination->next = $one;
            $three->next = $oldNext;

            // And move to the next cup
            $current = $current->next;
        }

        return $cupsByLabel[1];
    }

    /**
     * Runtime for part two would have been around 6,5 days.
     *
     * @return string
     */
    private function part1Original(): string
    {

        $cups = $this->cups;
        $max = max($this->cups);
        for ($i = 0; $i < 100; $i++) {
            $pickedUp = array_slice($cups, 1, 3);
            $dest = $cups[0];
            do {
                $dest--;
                if ($dest === 0) {
                    $dest = $max;
                }
            } while (in_array($dest, $pickedUp));
            $new = array_merge([$cups[0]], array_slice($cups, 4));
            $destinationOffset = array_search($dest, $new) + 1;
            $new = array_merge(array_slice($new, 0, $destinationOffset), $pickedUp, array_slice($new, $destinationOffset));
            $cups = array_merge(array_slice($new, 1), [$new[0]]);
        }
        $start = array_search(1, $cups);
        $result = '';
        $count = count($this->cups);
        for ($i = 1; $i < $count; $i++) {
            $result .= $cups[($start + $i) % $count];
        }

        return $result;
    }
}

